src/Pos/Txp/Logic/Global.hs

```diff
@@ -10,130 +10,119 @@ module Pos.Txp.Logic.Global
-           case verdict of
-               Right _     -> pass
-               Left errors ->
-                   assertionFailed $
-                   "txVerifyBlocks failed in txApplyBlocks call: " <> errors
-    verdict <- runTxpAction $
+    inAssertMode $ do
+        verdict <- runExceptT $ txVerifyBlocks blocks
+        whenLeft verdict $ assertionFailed .
+            sformat ("txVerifyBlocks failed in txApplyBlocks call: "%build)
+    txpModifierToBatch . snd <$> runToilAction
 #ifdef WITH_EXPLORER
-                   mapM (uncurry (applyTxp curTime) . blundToAuxNUndoWHash) blunds
+        (mapM (uncurry (applyToil curTime) . blundToAuxNUndoWHash) blunds)
 #else
-                   mapM (applyTxp . blundToAuxNUndo) blunds
+        (mapM (applyToil . blundToAuxNUndo) blunds)
 #endif
-    case verdict of
-        Left er           -> throwEx $ pretty er
-        Right (_, txpMod) ->
-            pure $ txpModifierToBatch txpMod
-  where
-    throwEx = throwM . TxpInternalError . (<>) "txApplyBlocks failed: "
 
 -- | Rollback chain of blocks.
 txRollbackBlocks
     :: (WithLogger m, MonadDB m)
     => NewestFirst NE (Blund ssc) -> m SomeBatchOp
-txRollbackBlocks blunds = do
-    verdict <- runTxpAction $
-                   mapM (rollbackTxp . blundToAuxNUndo) $ blunds
-    case verdict of
-        Left er           -> throwEx $ pretty er
-        Right (_, txpMod) -> pure $ txpModifierToBatch txpMod
-  where
-    throwEx = throwM . TxpInternalError . (<>) "txRollbackBlocks failed: "
+txRollbackBlocks blunds =
+    txpModifierToBatch . snd <$>
+    runToilAction (mapM (rollbackToil . blundToAuxNUndo) blunds)
 
 ----------------------------------------------------------------------------
 -- Helpers
 ----------------------------------------------------------------------------
 
--- Convert TxpModifier to batch of database operations.
-txpModifierToBatch :: TxpModifier -> SomeBatchOp
-txpModifierToBatch (TxpModifier um
-                       (BalancesView (HM.toList -> stakes) total)
+-- Convert ToilModifier to batch of database operations.
+txpModifierToBatch :: ToilModifier -> SomeBatchOp
+txpModifierToBatch (ToilModifier um (BalancesView (HM.toList -> stakes) total)
 #ifdef WITH_EXPLORER
-                       (MemPool _ _ em) _) =
+                       (MemPool _ _ em)
 #else
-                       _ _) =
+                       _
 #endif
-    SomeBatchOp
-        [ SomeBatchOp $
-          map GS.DelTxIn (MM.deletions um) ++
-          map (uncurry GS.AddTxOut) (MM.insertions um)
-        , SomeBatchOp $
-          map (uncurry GS.PutFtsStake) stakes ++ [GS.PutFtsSum total]
+                    _) =
+#ifdef WITH_EXPLORER
+    SomeBatchOp [SomeBatchOp utxoOps, SomeBatchOp balancesOps, SomeBatchOp explorerOps]
+#else
+    SomeBatchOp [SomeBatchOp utxoOps, SomeBatchOp balancesOps]
+#endif
+  where
+    utxoOps =
+        map GS.DelTxIn (MM.deletions um) ++
+        map (uncurry GS.AddTxOut) (MM.insertions um)
+    balancesOps =
```

Comment:
I spent 15 seconds deciphering this `maybe`, so maybe it'd be better to rewrite it as a `case` even if it would lead to some repetition?